### PR TITLE
Make return type implicitly None for type checked __init__ and __init_subclass__

### PR DESCRIPTION
**Pull request by onlined**
Implements https://github.com/python/mypy/issues/604#issuecomment-348525995.

### CODE REVIEW

File: mypy/semanal.py [hunk 1]
```diff
@@ -407,6 +406,10 @@ def _visit_func_def(self, defn: FuncDef) -> None:
                             add_symbol = False
                     if add_symbol:
                         self.type.names[defn.name()] = SymbolTableNode(MDEF, defn)
+                if defn.type is not None and defn.name() in ('__init__', '__init_subclass__'):
```
Comment (gvanrossum) [inline mypy/semanal.py]: Are you sure this works with overloads and decorated methods? Please add tests to verify that.

Also perhaps it should only work inside a class definition?

File: mypy/semanal.py [hunk 2]
```diff
@@ -68,8 +68,7 @@
 from mypy.types import (
     FunctionLike, UnboundType, TypeVarDef, TupleType, UnionType, StarType, function_type,
     CallableType, Overloaded, Instance, Type, AnyType,
-    TypeTranslator, TypeOfAny, TypeType,
-)
+    TypeTranslator, TypeOfAny, TypeType, NoneTyp)
```
Comment (gvanrossum) [inline mypy/semanal.py]: Please leave the `)` on a line by itself and add a trailing comma, preserving the original formatting.

File: mypy/semanal.py (hunk 1 above)
Comment (ilevkivskyi) [reply mypy/semanal.py]: I think one needs similar check at least in `visit_overloaded_func_def`.

Comment (onlined) [reply mypy/semanal.py]: @gvanrossum It works only inside a class definition.
@gvanrossum  @ilevkivskyi You are right, I will find a way to extend behavior.

**Comment by gvanrossum** [comment]
Can you also add a test for overloading?

**Comment by onlined** [comment]
I'm on it, but I can say that this change works with overloading too.

**Comment by onlined** [comment]
@gvanrossum It may be good to add tests for explicit `Any` return type, what do you think about it?

**Comment by gvanrossum** [comment]
> @gvanrossum It may be good to add tests for explicit `Any` return type, what do you think about it?

Yes, please.

**Comment by onlined** [comment]
I think this PR is ready to merge.

**Comment by gvanrossum** [comment]
Sorry, I got busy with other stuff. Will review later today.

**Comment by gvanrossum** [comment]
Done (sorry for the delay). Thanks for fixing this long-standing issue!

**Pull request merged**